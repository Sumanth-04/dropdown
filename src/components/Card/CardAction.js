import React, {Component} from 'react';
import PropTypes from 'prop-types';

class CardAction extends Component{
	static propTypes = {
		action: PropTypes.func.isRequired,
		actionStyles: PropTypes.object,
		className: PropTypes.string
	  };
	
	constructor(props){
		super(props);
	}

	render(){
		return (
			<button className={`rd-cardAction ${this.props.className}`} style={{...this.props.actionStyles}} onClick={() => this.props.action()}>{this.props.actionText}</button>
		)
	}
}

export default CardAction;