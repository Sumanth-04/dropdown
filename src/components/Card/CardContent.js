import React, {Component} from 'react';
import PropTypes from 'prop-types';

class CardContent extends Component{
	static propTypes = {
		text: PropTypes.string.isRequired,
		textStyles: PropTypes.object,
		className: PropTypes.string,
	  };
	
	constructor(props){
		super(props);
	}

	render(){
		return (
			<div className={`rd-cardContent ${this.props.className}`} style={{...this.props.textStyles}}>{this.props.text}</div>
		)
	}
}

export default CardContent;