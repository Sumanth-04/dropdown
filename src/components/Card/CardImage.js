import React, {Component, createRef} from 'react';
import PropTypes from 'prop-types';

const placeHolder =
  "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII=";

let observer;
let didCancel = false;

class CardImage extends Component{
	static propTypes = {
		src: PropTypes.string.isRequired,
		alt: PropTypes.string,
		imageStyle: PropTypes.object,
		className: PropTypes.string,
	  };
	
	constructor(props){
		super(props);
		this.imageRef = createRef();
		this.state = {
			imageSrc : placeHolder
		}
	}

	componentDidMount(){
		didCancel = false;
		if (this.imageRef && this.state.imageSrc !== this.props.src) {
		  if (IntersectionObserver) {
			observer = new IntersectionObserver(
			  entries => {
				entries.forEach(entry => {
				  if (
					!didCancel &&
					(entry.intersectionRatio > 0 || entry.isIntersecting)
				  ) {
					this.setState({ imageSrc : this.props.src });
					//this.imageRef.current ? observer.unobserve(this.imageRef.current) : null;
				  }
				});
			  },
			  {
				threshold: 0.01,
				rootMargin: "75%"
			  }
			);
			observer.observe(this.imageRef.current);
		  } else {
			// Old browsers fallback
			this.setState({ imageSrc : this.props.src });
		  }
		} else {
			// Old browsers fallback
			this.setState({ imageSrc : this.props.src });
		}
	}

	componentWillUnmount(){
		didCancel = true;
        // on component cleanup, we remove the listner
        if (observer && observer.unobserve) {
          observer.unobserve(this.imageRef.current);
        }
	}

	onLoad = event => {
		event.target.classList.add("loaded");
	  };
 
	 onError = event => {
		event.target.classList.add("has-error");
	  };

	render(){
		return (
			<img 
				ref={this.imageRef}
				className={`${this.props.className}`} 
				style={{...this.props.imageStyle}} 
				alt={this.props.alt}
				src={this.state.imageSrc} 
				onLoad={this.onLoad}
				onError={this.onError}
			/>
		)
	}
}

export default CardImage;