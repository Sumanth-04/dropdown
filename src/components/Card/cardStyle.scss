@import '../keyframes';


$placeHolder : 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII=';

.loader {
	overflow: hidden;
	background: white;
	border-radius: .25rem;
	box-shadow: 
	  0 15px 30px 0 rgba(0,0,0,0.05),
	  0 5px 15px 0 rgba(0,0,0,.05);
	transition: ease box-shadow 0.3s;
	&:hover {
	 box-shadow: 
	  0 15px 60px 0 rgba(0,0,0,0.08),
	  0 5px 25px 0 rgba(0,0,0,.08);
	}
  }
  

  .loader_screen {
	margin: 0;
	padding: 0;
	overflow: hidden;
  
	img {
	  height: auto;
	}
  }
  
  
  .loading {
	position: relative;
	background-color: #E2E2E2;
  
	&.card-image {
	  border-radius: 0;
	}
  
	&::after {
	  display: block;
	  content: '';
	  position: absolute;
	  width: 100%;
	  height: 100%;
	  transform: translateX(-100%);
	  background: linear-gradient(90deg, transparent, rgba(255, 255, 255, .2), transparent);
	  animation: loading 1.5s infinite;
	}
  }
  
.rd-cardImage{
	display: block;
  
  // I use utilitary classes instead of props to avoid style regenerating
  &.loaded:not(.has-error) {
    animation: loaded 300ms ease-in-out;
  }

  &.has-error {
    // fallback to placeholder image on error
    content: url($placeHolder);
  }
}