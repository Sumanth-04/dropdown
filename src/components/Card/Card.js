import React, {Component} from 'react';
import './cardStyle.scss';
import PropTypes from 'prop-types';

class Card extends Component{
	static propTypes = {
		cardStyles: PropTypes.object,
		className: PropTypes.string,
		};
		
	constructor(props){
		super(props);
	}

	render(){
		const {
			className,
			cardStyles,
			children,
			loading,
			loader
		} = this.props;

		return (
			<div className={`rd-card ${className}`} style={{...cardStyles}}>
				{loading ? loader : children}
			</div>
		)
	}
}

export default Card;