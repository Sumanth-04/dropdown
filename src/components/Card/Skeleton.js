import React, {Component} from 'react';
import './cardStyle.scss';

const Skeleton = ({width, height}) => { 
	return(
		<section class="loader" style={{width:width}}>
			<figure class="loader_screen loading" style={{height:height}}></figure>
	  	</section>
	)
}

export default Skeleton;