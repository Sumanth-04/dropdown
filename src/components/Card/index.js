import Card from './Card';
import CardAction from './CardAction';
import CardContent from './CardContent';
import CardImage from './CardImage';

//import '../../../static/style.scss';
Card.CardAction = CardAction;
Card.CardContent = CardContent;
Card.CardImage = CardImage;

export default Card;