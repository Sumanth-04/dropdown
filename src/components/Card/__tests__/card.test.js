import React from 'react';
import renderer from 'react-test-renderer';
import Card from '../index'

const CardAction = Card.CardAction;
const CardContent = Card.CardContent;
const CardImage = Card.CardImage;

const hi = () => { console.log("hi") }

it('Card Renders with image, content, actions', () => {
	const tree = renderer
		.create(
			<Card>
				<CardImage
					alt={"example"}
					src={'https://os.alipayobjects.com/rmsportal/QBnOOoLaAfKPirc.png'}
				/>
				<CardContent
					text={'Hello'}
					textStyles={{
						color: 'red'
					}}
				/>
				<CardAction
					actionStyles={{
						background: 'red'
					}}
					action={hi}
					actionText={"Action!"}
				/>
			</Card>
		)
		.toJSON();
		expect(tree).toMatchSnapshot();
});