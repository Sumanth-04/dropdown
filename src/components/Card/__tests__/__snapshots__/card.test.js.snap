// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Renders with image, content, actions 1`] = `
<div>
  <img
    alt="example"
    src="https://os.alipayobjects.com/rmsportal/QBnOOoLaAfKPirc.png"
  />
  <div
    style={
      Object {
        "color": "red",
      }
    }
  >
    Hello
  </div>
  <button
    onClick={[Function]}
    style={
      Object {
        "background": "red",
      }
    }
  >
    Action!
  </button>
</div>
`;
